File: datasets/python_source-code/Multiple-A1/organic4_binary_search.py
"""
Binary Search - Submission by Louise
Approach: Recursive implementation with default lo/hi arguments so
callers only need to pass the array and the target.
"""


def find(arr, target, lo=0, hi=None):
    if hi is None:
        hi = len(arr) - 1

    if lo > hi:
        return -1

    mid = (lo + hi) // 2
    current = arr[mid]

    if current == target:
        return mid
    if current > target:
        return find(arr, target, lo, mid - 1)
    return find(arr, target, mid + 1, hi)


if __name__ == "__main__":
    ARR = [2, 5, 8, 12, 16, 23, 38, 45, 56, 72, 91]
    TARGET = 23
    print(f"Target located at: {find(ARR, TARGET)}")

File: server/datasets/java_source-code/Multiple-A1/organic4_binary_search.java
/*
 * Binary Search - Submission by Louise
 * Approach: Recursive implementation with helper method overloading
 * so callers only need to pass the array and target.
 */

class BinarySearch {
    public static int find(int[] arr, int target) {
        return find(arr, target, 0, arr.length - 1);
    }

    public static int find(int[] arr, int target, int lo, int hi) {
        if (lo > hi) {
            return -1;
        }

        int mid = (lo + hi) / 2;
        int current = arr[mid];

        if (current == target) {
            return mid;
        }
        if (current > target) {
            return find(arr, target, lo, mid - 1);
        }
        return find(arr, target, mid + 1, hi);
    }

    public static void main(String[] args) {
        int[] arr = {2, 5, 8, 12, 16, 23, 38, 45, 56, 72, 91};
        int target = 23;
        System.out.println("Target located at: " + find(arr, target));
    }
}
